functions/go/set-labels/transformer/labels.go

```diff
@@ -0,0 +1,168 @@
+}
+
+func SetLabels(rl *fn.ResourceList) (bool, error) {
+	transformer := LabelTransformer{}
+	if err := transformer.Config(rl.FunctionConfig); err != nil {
+		rl.Results = append(rl.Results, fn.ErrorResult(err))
+		return false, nil
+	}
+	if err := transformer.Transform(rl.Items); err != nil {
+		rl.Results = append(rl.Results, fn.ErrorResult(err))
+		return false, nil
+	}
+	return true, nil
+}
+
+type LabelTransformer struct {
+	// Desired labels
+	NewLabels map[string]string `json:"labels,omitempty" yaml:"labels,omitempty"`
+	// FieldSpecs is deprecated, please use AdditionalLabelFields instead.
+	FieldSpecs []FieldSpec `json:"fieldSpecs,omitempty" yaml:"fieldSpecs,omitempty"`
+	// AdditionalLabelFields is used to specify additional fields to add labels.
+	AdditionalLabelFields []FieldSpec `json:"additionalLabelFields,omitempty" yaml:"additionalLabelFields,omitempty"`
+	// Results is used to track labels that have been applied
+	Results fn.Results
+}
+
+// Config parse the functionConfig kubeObject to the fields in the LabelTransformer
+func (p *LabelTransformer) Config(o *fn.KubeObject) error {
+	// parse labels to NewLabels
+	switch {
+	case o.IsEmpty():
+		return fmt.Errorf("FunctionConfig is missing. Expect `ConfigMap` or `SetNamespace`")
+	case o.IsGVK("", "", "ConfigMap"):
+		p.NewLabels = o.NestedStringMapOrDie("data")
+		if len(p.NewLabels) == 0 {
+			return fmt.Errorf("`data` should not be empty")
+		}
+	case o.IsGVK(fnConfigGroup, fnConfigAPIVersion, legacyFnConfigKind):
+		fallthrough
+	case o.IsGVK(fnConfigGroup, fnConfigAPIVersion, fnConfigKind):
+		p.NewLabels = o.NestedStringMapOrDie("labels")
+		if len(p.NewLabels) == 0 {
+			return fmt.Errorf("failed to configure function: input label list cannot be empty")
+		}
+	default:
+		return fmt.Errorf("unknown functionConfig Kind=%v ApiVersion=%v, expect `%v` or `ConfigMap`",
+			o.GetKind(), o.GetAPIVersion(), fnConfigKind)
+	}
+	// add default fields
+	if err := p.addDefaultLabelFields(); err != nil {
+		return err
+	}
+	// add additional fields
+	if o.IsGVK(fnConfigGroup, fnConfigAPIVersion, fnConfigKind) {
+		arr, exist, err := o.NestedSlice("additionalLabelFields")
+		if err != nil {
+			return fmt.Errorf("`additionalLabelFields` format is wrong")
+		}
+		if exist {
+			var addFields []FieldSpec
+			for _, sub := range arr {
+				var curField FieldSpec
+				if err := sub.As(&curField); err != nil {
+					return err
+				}
+				addFields = append(addFields, curField)
+			}
+			p.AdditionalLabelFields = append(p.AdditionalLabelFields, addFields...)
+		}
+	}
+	return nil
+}
+
+func (p *LabelTransformer) Transform(objects fn.KubeObjects) error {
+	for _, o := range objects {
+		for _, sp := range p.AdditionalLabelFields {
+			if (sp.Group == "" && sp.Version == "" && sp.Kind == "") || o.IsGVK(sp.Group, sp.Version, sp.Kind) {
+				newResult := fn.Result{
+					Message:     "Replace labels",
+					Severity:    "INFO",
+					ResourceRef: nil, // TODO: initialize in kyaml
```

Comment:
Do you want to do this in this PR or in a followup PR.

Reply:
I reconsider this and now I think since the E2E test's result format doesn't contains the RecourceRef. So I can simply set it to be nil. Is it OK for me to leave `ResourceRef` to `nil`?